### DISCUSSION TITLE
Windows port cmake build

### DISCUSSION THREAD

**Discussion by kdschlosser** [micropython/micropython#14465] [Unix & Windows]
I am wanting to see if there would be interest in adding support to compile on windows using make and cmake. There is a windows binary of make that can be used to keep with how the rest of the ports are compiled. I am talking about a native windows build using MSVC and not mingw or wsl. It would require visual studio be installed or the Visual Studio build tools. I have written a python script that is able to locate and setup a proper build environment for Windows which can be used to help things out. cmake is available as part of Visual Studio so no problem there either.

This could be further expanded to compile for stm32, the rp2 and other MCU's as well. There are requirements that would need to be done to make it work. Thing like installing the arm toolkit for windows if compiling for stm32 or rp2 and the esp-idf toolkit if compiling for esp32. This is able to be done now that all of the various pieces have been made available. 

arm toolkit
https://developer.arm.com/downloads/-/gnu-rm

esp-idf toolkit
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/windows-setup.html

gnu make for windows
https://gnuwin32.sourceforge.net/packages/make.htm

sed for windows
https://gnuwin32.sourceforge.net/packages/sed.htm

That's to get the basic idea that everything is available.

Now make doe4sn't know anything about MSVC which is why cmake would be used. Make could check the OS and then launch cmake provided there is windows support in the current cmake files in the ports that use cmake and cmake files would need to be added to the ports that use make.

As a start off point it would be nice if the windows port was able to be compiled with a user c module and it should be able to be compiled without using msbuild. I don't know cmake or make for that matter makefile well enough to write those files. I do know msvc. If I could get together with someone that knows cmake we could get a build system going to support compiling natively on windows across all ports. After all Windows does have 72% market share for PC's and not having the ability to compile on Windows without needing to jump through hoops with cross compiling would be a nice addition to MicroPython. It would make it a lot easier for a lot of folks to be able to do that. 

The tools are available to get it done....

Anyone that has interest in having this added speak up and let the devs know by responding to to post. If there is anyone that has knowledge that would help get it off the ground and you want to share here is a good place to start. If you want to lend a hand with the coding, make it known here...

I was going to open an issue for this and then I though it would be a better place to see if there was any interest in it and what the help situation looked like as well. 

I do have everything compiling properly on Linux and also on macOS and the last piece is Windows.

**Comment by stinos** [comment]
I'd personally stay away from `make` for the msvc port, mainly because it's not super standard and we have already 2 ways to build working executables for Windows. As such I don't see any benefit of maintaining more build tool scripts, but I might be missing something there - what would be the key reasons why such make-based build would be beneficial (and/or easier than what we have now)? And/or for having make launch cmake?

Same question for cmake;  I'm not sure if it brings much to the table?

Unless the end goal is to have a single cmake-based build for all PC/Mac based ports (i.e. unify the unix and windows ports into one cmake 'PC' port). While that makes it a bit harder and less convenient to 'just build' on windows I do think it would be pretty nice overall..



> I have written a python script that is able to locate and setup a proper build environment for Windows which can be used to help things out.

Unless sticking to > VS2022 this is notoriously finicky with the various different versions of build tools and Windows SDK. If anything, this should imo be offloaded to an external tried&tested tool (or maybe that's what you did and your script just calls into that?). But then it would likely be easier to go for cmake instead because it should always work in that regard.



> This could be further expanded to compile for stm32, the rp2 and other MCU's as well. 

I'd think this would be a more interesting goal: 'one-click' builds on windows for other platforms. As in: it'll download and/or update the build tools for you unless you have them already then build. But that's more of a job for a basic script (Python or Powershell) than make/cmake.

**Reply by kdschlosser** [reply]
and what are the 2 working builds for Windows? the VS sure doesn't work. If a user c module is not able to be compiled and a user is not able to freeze modules into it then it's incomplete/doesn't work properly.

what is the other way to build it?


I am working on what could be called a "one click" build system for MicroPython. its a single entry point that handles collecting any dependencies and running the build for the different MCU's on different platforms. This is where I am having an issue. I am not able to compile MicroPython for the esp32, rp2, stm32, or any other port on Windows nor am I am to compile MicroPython to run on Windows. I have gotten everything going for macOS and Linux and the one OS that is on the largest number of PC's in the world by far has zero support. makefile is a dead stick as far as build systems go because it has one serious flaw. It's not cross platform. cmake seems like it is becoming more commonplace and when you couple it with ninja you gain the ability to compile using multiple cores which takes care of the one thing that cmake was lacking in.

**Reply by andrewleech** [reply]
Personally, my preferred compile system for all ports is docker containers - they're 100% repeatable and allow all dependencies to be self contained and distributed reliably.
It means you can trivially jump between different version toolchains in and between ports too.
They also don't require replacing / replicating any of the existing build scripts which is important.

I've worked on getting make based builds on native windows on past projects and it's always a nightmare keeping it working. There's just such huge variation between different people's computer setups, dealing with path issues and conflicting tools getting in the way.
I'm not sure if anything's changed but when I previously tried the native build of the make tool it was weirdly slow, like would take minutes to run a makefile that only took seconds in Linux, not sure why. 
Also, most people who write and maintain makefiles think nothing of using standard *nix command line tools in the meantime l makefile, like grep, sed, cp and even bash filtering. These then break badly if you try to run on native windows. Its just harder to maintain multiple platforms, especially if the person writing the makefile doesn't own a windows PC (like many of the maintainers).

I've been using the mingw windows build for years, built in a docker container automatically in CI, it's fast and reliable.
I also know the visual studio build is actively maintained by @stinos, if there actually are features missing I'd prefer they just get fixed rather than trying to replace the lot. 

Side note; I've used cmake a bit. I _loath_ it. Every time I run into it again it takes _so_ much longer for me to figure out how to get it to just work than a makefile. cmake is a whole language and programming environment in itself and it's a big ecosystem which makes it hard to learn enough to get productive with it. While makefiles also have a learning curve for sure, it's really just a basic script with variables and build script snippets at the end of the day.

I've spoken with @mattytrentini in the past about making it easy (one click) for people to build with docker containers, the top level tool could be given just a board name and it'd simply find the matching port, pull the matching docker container then run the existing make command in the new container. 
The only dependency needed at this point would be docker (which has clear install instructions for all major platforms) and whatever the tool itself needs (likely python but could be bash/PowerShell)

**Reply by kdschlosser** [reply]
docker is not a build system, it is a sandbox virtual environment that you compile in. docker doesn't do any compiling or instruct any compiles or anything of the sort. It takes a snapshot prior to compiling so the environment doesn't need to get set up over and over again. 

The other issue is docker is not cross platform either. It doesn't run on Windows. It uses a Linux virtual machine to run. anything ARM related is only available in the cloud... 

Docker is not what is needed.

**Reply by kdschlosser** [reply]
for windows it would be a cmake based build system the only thing make would do is call cmake. It is only to keep the user entry point all the same across the board. Just like compiling for an ESP32. make is used only for some menial tasks and everything gets done by cmake

**Reply by andrewleech** [reply]
> docker is not a build system, it is a sandbox virtual environment that you compile in. docker doesn't do any compiling or instruct any compiles or anything of the sort. It takes a snapshot prior to compiling so the environment doesn't need to get set up over and over again.

Yep that's the entire point. There's already a build system in place that's reliable and well tested. Docker makes this trivially easy to run on any system. It's a compliment to the current build system, without needing to replicate anything.

It runs _very_ well on windows systems, I use it every day for work and micropython hacking. Yes, behind the scenes it uses WSL and a bunch of OS layers to make it happen, but these things are all managed automatically and again, the documentation and maintenance of these systems is done by other large, well funded companies so we wouldn't need to do it ourselves - just point users to the docker install docs and then come back here to run "./build <my board>"

arm builds are also quite easy to cross compile in docker, faster than many native systems :-D

**Reply by andrewleech** [reply]
> for windows it would be a cmake based build system the only thing make would do is call cmake. It is only to keep the user entry point all the same across the board. Just like compiling for an ESP32. make is used only for some menial tasks and everything gets done by cmake

cmake is _only_ used on esp and rp2 ports because their manufacturer libraries require it, these are by far the minority, especially when the cmake setup for the two of these is already quite different (due to said manufacturer build setups being different). 

All other ports are pure Makefile based. Adding cmake to any of these would increase the mainetenance burden of them all. I for one am strongly against this, and duplication / multiplication of build systems makes it harder to add or maintain features what touch the build system. I would much prefer to remove cmake from the existing ports like circuitpython does.

**Reply by kdschlosser** [reply]
https://stackoverflow.com/questions/29650170/can-docker-help-build-executable-that-work-in-different-platform)

Docker doesn't work on Windows OR macOS. so you would NOT be able to compile the Unix or Windows port on docker and have the resulting compiled MicroPython binary run on Windows and macOS. It won't work.....


The reason why these things need to be compiled on the OS they are going to be running on is because of user c modules. As an example.. 

https://github.com/kdschlosser/lvgl_micropython

That repo is a user c module that allows access to the graphics framework LVGL. Part of that is using SDL and being able to link to SDL. MSVC is what needs to do the compiling on Windows and for the Unix port being compiled on macOS clang needs to be used it has to be compiled on macOS.

Docker is Linux ONLY. You cannot have docker create a Windows Image and docker will not run on a Windows machine without using a virtual machine that is running Linux. The same thing applies to macOS

**Reply by kdschlosser** [reply]
I am also aware that cmake is only used for the RP2 and the ESP32 ports. There is a reason why and that is because the companies that make those MCUs wanted people to be able to compile on Windows as well as the posix OS's. make is not cross platform. In fact espressif used to be compiled using make. It was replaced with a build framework that is cross platform and that is cmake. the RP2 more than likely also used cmake for the same reason. They didn't want to loose out on the sales because their product is not able to compile on 72% of the PC's in the world.

cross platform means it is able to be compiled and run across multiple OS's and machine architectures. It also needs to be able to compile in full which means user c modules also need to be able to compile for all of the different ports.

**Reply by andrewleech** [reply]
> Docker doesn't work on Windows OR macOS. so you would NOT be able to compile the Unix or Windows port on docker and have the resulting compiled MicroPython binary run on Windows and macOS. It won't work.....
> 
> The reason why these things need to be compiled on the OS they are going to be running on is because of user c modules. As an example..

I already compile native windows and mac builds in micropython / mpy-cross inside docker containers, it works just fine, been doing it for years. 
It's using standard practice cross-compiling.

**Reply by andrewleech** [reply]
> not able to compile on 72% of the PC's in the world.

or depending on what question you ask...

> 96.3% of the top one million web servers are running Linux. (ZDNet)

Most companies use CI for building products, which are run on servers of some kind, the overwhelming majority of which are linux. Even the embedded SW developers at work with windows desktops _all_ use linux in WSL / docker / vm for development, because windows lost that race a few decades ago in the SW dev space.

**Reply by andrewleech** [reply]
Sorry I _really_ didn't mean / want to start a fight over this! 

I just work in this every day professionally where we've been rolling out docker based dev / build environments to great success (vscode devcontainer & matching CI), so wanted to share some of my experiences in this. 

I really tried hard to make some of my past projects compile seamlessly on every platform with cmake / make / other tools and it was always a fight to get it working in the first place and even harder to keep it maintained and working. Docker has just simplified things so so much for me.

The goals of cmake are good, but the actuality is too overcomplicated in my opinion. But, that is just my professional opinion, nothing more.

**Reply by kdschlosser** [reply]
You are not hearing me. YOU CANNOT COMPILE A USER C MODULE WITH THE WINDOWS BUILD IT DOESN'T WORK I CANNOT CROSS COMPILE BECAUSE OF USING SDL IT HAS TO BE COMPILED USING MSVC ON WINDOWS. NO VM THAT RUNS LINUX OR LINUX SUBSYSTEM CAN BE USED!!!!

The build system will fail if it needs to compile on Windows when using a user C module. cross compiling the windows port is NOT compiling on windows. It is compiling in unix for windows. There is a HUGE difference.
docker containers DO NOT run on windows. it's that simple. cross compiling using docker containers is running inside a linux VM. it is not compiling on windows.

I did not make it this way, but I am telling you that it doesn't work. If something is made to run on windows it should be compiled using MSVC because of the CLR. Windows gets to be a real fickle thing when it comes to that. While it may work cross compiling it as soon as something like a user c module needs to be added and that module needs to access some portion of Windows the build will fall flat on it's face.

**Reply by kdschlosser** [reply]
If the entire build system was unified into cmake it would make it vastly simpler and easier in every which way. There would be no need to have .mk files and then .cmake files that for all intents and purposes does the same thing. Most of the build system is already coded for cmake. it's the ports that aren't. pretty much everything else is.

**Reply by kdschlosser** [reply]
go look at the build system you will see that the majority of it is already coded for cmake.

**Reply by stinos** [reply]
> Personally, my preferred compile system for all ports is docker containers

@andrewleech right, didn't you or Matt have a repository with docker configs or something like that?

> Side note; I've used cmake a bit. I loath it.

I get what you mean, but on the other hand: I have used both, but not repeatedly nor enough to get to actually know them by heart, and then the key difference for me is: with cmake I at least get the idea that I know what I'm doing after a couple of minutes. Whereas with make I first have to decipher its archaic syntax :) As an actual argument for/against that doesn't really count though..

**Reply by andrewleech** [reply]
> @andrewleech right, didn't you or Matt have a repository with docker configs or something like that?

Yep quite a few years ago I got a CI system building all boards and all ports, including windows and Mac Unix builds, all in docker containers on Gitlab. This was before GitHub actions existed, once that started being used here I stopped maintaining my repo. But the docker containers all still work fine to build all the ports.
I recently updated my Mac docker cross compiler for mpy-cross to add support for the Apple silicon / arm machine too.
When time permits I'd like to get a wiki page set to to describe how these can be used easily (without any changes to micropython codebase)

**Reply by andrewleech** [reply]
@kdschlosser I hear you clearly, however I've been trying to calmly share my experience with you of what does already work with micropython, without new build systems / practices needing to be written by an unknown cmake expert. 

I already build user C module on Unix / windows port with make, cross compiled with mingw in a docker container. 
In the past I've also used SDL and many other native windows libraries with the same mingw cross compiler on Linux build machines, it's a standard practice in the professional sw industry. I even sign mingw compiled windows applications with authenticode certificates as work (using Linux tools) that we then run and distribute on windows. This all works already. .

I'm well aware there are already duplicated build scripts throughout the codebase for make and cmake, you will see commits from me on many of them in the git history. 
The cmake fragments were ported from the original makefile ones just to support esp and rp2 ports. Makefile is still considered the original and primary build system by the maintainers I speak to, cmake is begrudgingly updated to match when needed.

**Comment by stinos** [comment]
> and what are the 2 working builds for Windows?

See https://github.com/micropython/micropython/blob/master/ports/windows/README.md: the msvc (msbuild as build system, so using the .vcxproj files) and mingw-w64 (make as build system, Makefile is very much like the one of all other ports).


> If a user c module is not able to be compiled and a user is not able to freeze modules into it then it's incomplete/doesn't work properly.

Unless I'm missing something both support freezing and the mingw-w64 one does support user C modules. Although the msvc one also supports variants and other customizations so it's not like it's impossible to get external code into it. If you think user C modules are a worthy addition for the msvc port then please raise an issue for that.

Like Andrew says: if one of the reasons for wanting a cmake-based build is the lack of a feature in a windows port, then implementing then implementing that feature in the current build system seems like the sane thing to do, instead of implementing something completely new from scratch.


> nor am I am to compile MicroPython to run on Windows.

Please raise an issue (or start a discussion) for this, because as far as I (and the CI builds) know everything is fine.


> I CANNOT CROSS COMPILE BECAUSE OF USING SDL

That make little sense. SDL should work on all major platforms? And if not cross-compilation then with mingw-w64? And that micropython port does user C modules (again, as far a I'm aware, last time I tried it did) so should be a viable option?


> If something is made to run on windows it should be compiled using MSVC because of the CLR

You probably mean CRT not CLR, but still that's not very accurate since other compilers (intel, gcc, ...) can use that.


> If the entire build system was unified into cmake it would make it vastly simpler and easier in every which way

From the user's point of view, probably. But as far as maintaining it goes I'm not sure it's going to actually be simpler let alone easier if you take everything into account (custom build steps, the documentation, having to learn a new tool, ...). And in any case it would be a major effort (just guessing this is going to go towards multiple man-months for all ports probably).
